src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,157 @@
+        }
+    }
+
+    private String outOfBoundsMsg(int index) {
+        return "Index: " + index + ", Size: " + size;
+    }
+
+    private Object[] growIfArrayFull(int minCapacity) {
+        int oldCapacity = elementData.length;
+        if (oldCapacity > 0 || elementData != DEFAULTCAPACITY_EMPTY_ELEMENTDATA) {
+            int newCapacity = (int)(oldCapacity * CAPACITY_GROWTH_RATIO);
+            Object[] elementDataNew = new Object[newCapacity];
+            System.arraycopy(elementData,0, elementDataNew, 0, elementData.length);
+            return elementDataNew;
+        }
+        return new Object[Math.max(DEFAULT_CAPACITY, minCapacity)];
+    }
 
+    private Object[] growIfArrayFull() {
+        return growIfArrayFull(size + 1);
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] elementDataNew = new Object[list.size()];
+        for (int i = 0; i < elementDataNew.length; i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isIndexInBounds(index)) {
+            return (T) elementData[index];
+        } else {
+            throw new ArrayListIndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+    }
+
+    private boolean isIndexInBounds(int index) {
+        return index <= size - 1 && index >= 0;
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (isIndexInBounds(index)) {
+            elementData[index] = value;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+        T element = (T) elementData[index];
+        removeElementByIndex(index);
+        return element;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T removedElement = null;
+        int index = NEGATIVE_INDEX_MARKER;
+        for (int i = 0; i < size; i++) {
+            if (elementData[i] != null && elementData[i].equals(element)) {
+                index = i;
+                removedElement = (T) elementData[i];
+                break;
+            }
+
+            if (element == null && elementData[i] == null) {
+                size--;
+                return null;
+            }
+        }
+
+        if (index == NEGATIVE_INDEX_MARKER) {
+            throw new NoSuchElementException();
+        }
+        removeElementByIndex(index);
+        return removedElement;
+    }
+
+    private void removeElementByIndex(int index) {
+        if (size - 1 > index) {
+            System.arraycopy(elementData, index + 1, elementData, index, size - 1 - index);
+        }
+        size--;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
```

Comment:
Don't begin class or method implementation with an empty line. Please remove the empty line before the 'isEmpty' method declaration.